### PR TITLE
Implement log, append log, clear log commands

### PR DESCRIPTION
**Pull request by hyc17003**

### CODE REVIEW

File: src/main/java/seedu/address/model/ModelManager.java [hunk 1]
```diff
@@ -30,6 +30,8 @@ public class ModelManager implements Model {
     private final UserPrefs userPrefs;
     private final FilteredList<Person> filteredPersons;
     private final FilteredList<Person> loggedFilteredPersons;
+    private String lastExecutedCommandType = "unknown";
```
Comment (WZWren) [inline src/main/java/seedu/address/model/ModelManager.java:33]: What would this `String` do? It doesn't seem to be referenced anywhere else, and this doesn't seem relevant to the `log` command.

File: src/main/java/seedu/address/model/ModelManager.java [hunk 2]
```diff
@@ -30,6 +30,8 @@ public class ModelManager implements Model {
...
+    private FilteredList<Person> foundPersonsList;
```
Comment (WZWren) [inline src/main/java/seedu/address/model/ModelManager.java:34]: Perhaps it is better to switch to the `loggedFilteredPersons` instead? @WeeeHung has created a `loggedFilteredPersons` list on `Line 32` in his previous PR #107, and you may have missed the change during a synchronization merge in a previous commit.

**Review by WZWren — COMMENTED** [review]
I will perform a more formal review when I have the time, but here are some of the comments I have at a glance.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 1 above)
Comment (hyc17003) [reply src/main/java/seedu/address/model/ModelManager.java:33]: Good catch, I had added it when pursuing an alternative implementation, and it slipped my eye. It will be removed in the next commit.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 2 above)
Comment (hyc17003) [reply src/main/java/seedu/address/model/ModelManager.java:34]: `loggedFilteredPersons` refers to the currently-logged filtered persons list. `foundPersonsList` keeps track of the results of the latest `find` command, used to add the results to the logbook. I used it so `log` will only save results of `find` commands, instead of the entire list, which would happen if `filteredPersons` was used.

File: src/main/java/seedu/address/logic/commands/AppendLogCommand.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ */
+public class AppendLogCommand extends Command {
+
+    public static final String COMMAND_WORD = "alog";
+
+    public static final String COMMAND_WORD_ALIAS = "al";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + "or" + COMMAND_WORD_ALIAS
+            + ": Appends the results of the most recent FindCommand to the logger tab.\n"
+            + "Example: " + COMMAND_WORD + "or" + COMMAND_WORD_ALIAS;
```
Comment (nubnubyas) [inline src/main/java/seedu/address/logic/commands/AppendLogCommand.java:19]: I think would be good to split into 2 examples, it is to standardise with the format in other classes.
```suggestion
            + "Example 1: " + COMMAND_WORD + "\n";
            + "Example 2: " + COMMAND_WORD_ALIAS;
```

File: src/main/java/seedu/address/logic/commands/LogCommand.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ */
+public class LogCommand extends Command {
+
+    public static final String COMMAND_WORD = "log";
+
+    public static final String COMMAND_WORD_ALIAS = "lo";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + "or" + COMMAND_WORD_ALIAS
+            + ": Saves the results of the most recent FindCommand to the logger tab.\n"
+            + "Example: " + COMMAND_WORD + "or" + COMMAND_WORD_ALIAS;
```
Comment (nubnubyas) [inline src/main/java/seedu/address/logic/commands/LogCommand.java:18]: Same as the comments in `ClearLogCommand` and `AppendLogCommand`.
```suggestion
            + "Example 1: " + COMMAND_WORD + "\n"
            + "Example 2: " + COMMAND_WORD_ALIAS
```

File: src/main/java/seedu/address/logic/commands/ClearLogCommand.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ */
+public class ClearLogCommand extends Command {
+
+    public static final String COMMAND_WORD = "clog";
+
+    public static final String COMMAND_WORD_ALIAS = "cl";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + "or" + COMMAND_WORD_ALIAS
+            + ": Clears the entire logger tab.\n"
+            + "Example: " + COMMAND_WORD + "or" + COMMAND_WORD_ALIAS;
```
Comment (nubnubyas) [inline src/main/java/seedu/address/logic/commands/ClearLogCommand.java:19]: Same thoughts as the one `AppendLogCommand`, can make it 2 examples.
```suggestion
            + "Example 1: " + COMMAND_WORD + "\n"
            + "Example 2: " + COMMAND_WORD_ALIAS;
```

**Review by nubnubyas — COMMENTED** [review]
The implementation looks good to me. Just some small standardization for `MESSAGE_USAGE`. Do add more test cases for the new commands as the codecov is low.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 2 above)
Comment (WZWren) [reply src/main/java/seedu/address/model/ModelManager.java:34]: This is certainly a valid way to do this. However, I would like to point out that `FilteredList` is primarily a [JavaFX Collection](https://openjfx.io/javadoc/17/javafx.base/javafx/collections/transformation/FilteredList.html), designed to interact with the UI. Since the `foundPersonsList` collection does not interact with the UI, perhaps it may be better to use an alternative solution?

For example, the `FilteredList` instances are directly backed by the `getPersonList` method in the constructor of `ModelManager` (`Lines 47 to 49`). Notice that if you track the method calls during the invocation of `execute` of the `AddCommand` and `EditCommand`, the `FilteredList` instances are never directly modified ever again - changes are only made to `AddressBook`, and `FilteredList` only gets a `Predicate` to refresh it.

I believe this can be attributed to the lazy property of the `Observable` interface, which the `FilteredList` inherits - by some black box miracle, the lazily evaluated `Observable` property lets the UI update with the latest information in `AddressBook` after inducing a state-change in the relevant `FilteredList` instance.

In other words, you could potentially get the last `find` result from the `FilteredList` instance by invoking the `getPredicate()` command on `filteredPersons`, like so:

```
public void updateLoggedFilteredPersonsOverriding() {
    loggedFilteredPersons.setPredicate(filteredPersons.getPredicate());
}

public void updateLoggedFilteredPersonsAppending() {
    loggedFilteredPersons.setPredicate(loggedFilteredPersons
            .getPredicate()
            .or(filteredPersons.getPredicate())
    );
}
```

File: src/main/java/seedu/address/logic/commands/AppendLogCommand.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+            + "Example 2: " + COMMAND_WORD_ALIAS;
+
+    public static final String MESSAGE_SUCCESS = "Results of the FindCommand have been appended to the logger tab.";
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        // Ensure that there are results from the most recent FindCommand
+        if (model.getFoundPersonsList().isEmpty()) {
+            throw new CommandException(Messages.MESSAGE_EMPTY_FIND_RESULT);
+        }
```
Comment (WZWren) [inline src/main/java/seedu/address/logic/commands/AppendLogCommand.java:29]: Check this comment for details on `FilteredList` - I believe that this could potentially never be empty because of how the `ObservableList` is designed.

File: src/main/java/seedu/address/logic/commands/AppendLogCommand.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+        for (Person person : model.getFoundPersonsList()) {
+            model.getLogBook().addPerson(person);
+        }
```
Comment (WZWren) [inline src/main/java/seedu/address/logic/commands/AppendLogCommand.java:33]: In a similar vein to the code above, this loop may be causing `addPerson` to add duplicates to the `LogBook` instance, as `LogBook` never explicitly checks if the person exists and `getFoundPersonsList` is potentially looking iterating through the entire list of the `AddressBook` rather than just the filtered section.

File: src/main/java/seedu/address/logic/commands/LogCommand.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+            + "Example 2: " + COMMAND_WORD_ALIAS
+
+    public static final String MESSAGE_SUCCESS = "Results of the FindCommand have been saved to the logger tab.";
+
+    @Override
+    public CommandResult execute(Model model) throws CommandException {
+        // Ensure that there are results from the most recent FindCommand
+        if (model.getFoundPersonsList().isEmpty()) {
+            throw new CommandException(Messages.MESSAGE_EMPTY_FIND_RESULT);
+        }
```
Comment (WZWren) [inline src/main/java/seedu/address/logic/commands/LogCommand.java:28]: See my comment above.

File: src/main/java/seedu/address/logic/commands/LogCommand.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+        model.getLogBook().setPersons(model.getFoundPersonsList());
```
Comment (WZWren) [inline src/main/java/seedu/address/logic/commands/LogCommand.java:30]: In retrospect, we may have been too hasty in approving the `LogBook` implementation. This is likely setting the `LogBook` internal list to follow the `ObservableList` instance of `AddressBook`.

**Review by WZWren — COMMENTED** [review]
In addition to my comments below, do [reference the reply I have made](https://github.com/AY2324S1-CS2103T-T14-3/tp/pull/117/files#r1377418989) in the previous review.

Particularly, the previously merged `LogBook` may be redundant in its implementation, since the `Logger` section and the main list section both inherit the same `ObservableList` backed by `AddressBook`.

File: src/main/java/seedu/address/logic/commands/AppendLogCommand.java (hunk 6 above)
Comment (hyc17003) [reply src/main/java/seedu/address/logic/commands/AppendLogCommand.java]: Good point, have replaced it with `assert` for documentation.

File: src/main/java/seedu/address/model/ModelManager.java (hunk 2 above)
Comment (hyc17003) [reply src/main/java/seedu/address/model/ModelManager.java:33]: Discussed in real life - will merge first and refactor thereafter to address this bug.

File: src/main/java/seedu/address/logic/commands/AppendLogCommand.java (hunk 7 above)
Comment (hyc17003) [reply src/main/java/seedu/address/logic/commands/AppendLogCommand.java:50]: Duplicates are not added - but defensive programming will be helpful here, good point.

**Comment by WeeeHung** [comment]
LGTM. Team have decided to merge this for dry run demo.
